### PR TITLE
Support V4TrainingData

### PR DESCRIPTION
**Pull request by Cyanogenoid**
See https://github.com/LeelaChessZero/lc0/pull/722.
This is backwards compatible to loading V3TrainingData by setting root_q, best_q, root_d, root_q to 0.0 and not masking away illegal moves. Setting q_ratio in the interval [0, 1] (default 0) allows adjustment of how much q is mixed into z. In training, this PR changes nothing by default.

I started a training run on CCRL with v3 data and it hasn't broken at 1000 steps yet.

### CODE REVIEW

File: tf/chunkparser.py [hunk 1]
```diff
@@ -113,7 +115,7 @@ def init_structs(self):
         struct.Struct doesn't pickle, so it needs to be separately
         constructed in workers.
 
-        V3 Format (8276 bytes total)
+        V3 Format (8292 bytes total)
```
Comment (Tilps) [inline tf/chunkparser.py]: V4 format?

File: tf/decode_training.py [hunk 2]
```diff
@@ -2095,11 +2103,18 @@ def main(args):
...
+                if version == VERSION3:
+                    record_size = V3_BYTES
+                else:
+                    record_size = V4_BYTES
+                for i in range(0, len(chunkdata), record_size):
+                    ts = TrainingStep(4 if version == VERSION4 else 3)
+                    record = chunkdata[i:i+record_size]
+                    if chunkdata[0:4] == VERSION3:
+                        record += 16 * b'\x00'
+                    ts.display_v2_or_v3(i//record_size, record)
```
Comment (Tilps) [inline tf/decode_training.py]: function name seems stale.

File: tf/tfprocess.py [hunk 3]
```diff
@@ -95,18 +95,25 @@ def init_net(self, next_batch):
...
                                                     logits=self.y_conv)
         self.policy_loss = tf.reduce_mean(cross_entropy)
 
         # Loss on value head
+        q_ratio = self.cfg['training'].get('q_ratio', 0)
+        assert 0 <= q_ratio <= 1
+        target = self.q_ * q_ratio + self.z_ * (1 - q_ratio)
         self.mse_loss = \
-            tf.reduce_mean(tf.squared_difference(self.z_, self.z_conv))
+            tf.reduce_mean(tf.squared_difference(target, self.z_conv))
```
Comment (Tilps) [inline tf/tfprocess.py]: I assume these changes don't cause any significant change in training speed. (I would hope not, but after SWA I like to ask...)

Comment (Cyanogenoid) [reply tf/tfprocess.py]: It's just interpolating two tensors of size (batch_size, 1), should be completely negligible. Didn't seem slower when I ran this for 1k steps).

File: tf/decode_training.py (hunk 2 above)
Comment (Cyanogenoid) [reply tf/decode_training.py]: v2 support doesn't seem to exist in chunkparser anymore, so removing that branch entirely.

File: tf/chunkparser.py (hunk 1 above)
Comment (Cyanogenoid) [reply tf/chunkparser.py]: Yep.

File: tf/chunkparser.py [hunk 4]
```diff
@@ -310,20 +337,20 @@ def generate_fake_pos(self):
...
 
-    def v3_record(self, planes, i, probs, winner):
+    def v4_record(self, planes, i, probs, winner):
         pl = []
         for plane in planes:
             pl.append(np.packbits(plane))
         pl = np.array(pl).flatten().tobytes()
         pi = probs.tobytes()
-        return self.v3_struct.pack(VERSION, pi, pl, i[0], i[1], i[2], i[3], i[4], i[5], i[6], winner)
+        return self.v4_struct.pack(V4_VERSION, pi, pl, i[0], i[1], i[2], i[3], i[4], i[5], i[6], winner)
```
Comment (Tilps) [inline tf/chunkparser.py]: Shouldn't this have some more fields added to its packing?

Comment (Cyanogenoid) [reply tf/chunkparser.py]: Tests pass now.

File: tf/train.py [hunk 5]
```diff
@@ -133,7 +133,7 @@ def main(cmd):
     # Assumes average of 10 samples per test game.
     # For simplicity, testing can use the split batch size instead of total batch size.
     # This does not affect results, because test results are simple averages that are independent of batch size.
-    num_evals = num_test*10 // ChunkParser.BATCH_SIZE
+    num_evals = num_test//10 // ChunkParser.BATCH_SIZE
```
Comment (Tilps) [inline tf/train.py]: Please revert - then good to submit.

**Comment by Tilps** [comment]
This has broken some of the supplementary scripts that create a dummy data entry in order to start init_net - they need to be updated for the additional passed parameter.
